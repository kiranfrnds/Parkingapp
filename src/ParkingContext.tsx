import React, {useState } from "react";
import { createContext } from "react";

export interface ParkingProps {
    parkingSlots : any[],
    setParkingSlots: (parkingSlots:any) => void,
    stateId:any,
    setStateId:any
}

export const ParkingContext = createContext({} as ParkingProps )

export const ParkingContextProvider = ({children}: any) => {
    const [parkingSlots, setParkingSlots] = useState<any[]>([])
    const [stateId, setStateId] = useState<any>()

    return (
        <ParkingContext.Provider value={{
            parkingSlots,
            setParkingSlots,
            stateId,
            setStateId

        }} 
        >
            {children}
        </ParkingContext.Provider>
    )
}