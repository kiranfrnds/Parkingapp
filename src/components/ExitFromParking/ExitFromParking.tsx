import { Container, Box, Typography, Button } from "@mui/material";
import React, { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { ParkingContext } from "../../ParkingContext/ParkingContext";

const ExitFromParking = () => {
  const navigate = useNavigate();
  const { parkingSlots, stateId } = useContext(ParkingContext);

  const handlePayNow = async (carNumber: any) => {
    let index: any = parkingSlots?.findIndex(
      (item: any) => item.id === stateId
    );

    let regNumber = {
      carNumber: carNumber,
    };

    fetch("https://httpstat.us/200", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(regNumber),
    })
      .then((data) => {
        data.status === 200
          ? alert("Payment Successful")
          : alert("Payment Unsuccefull");
      })
      .catch((err) => alert(err));
    if (index !== "") {
      parkingSlots[index].carNumber = "";
      parkingSlots[index].available = true;
    }
    navigate("/slots");
  };

  let exitSlotDetails: any = parkingSlots.find(
    (car: any) => car.id === stateId
  );

  console.log(exitSlotDetails);

  let entryHours: any = exitSlotDetails?.time
    ? exitSlotDetails.time.getHours()
    : null;
  let entryMinutes = exitSlotDetails?.time.getMinutes();

  let currentHours = exitSlotDetails?.time.getHours() + entryHours;
  let currentMinutes = exitSlotDetails?.time.getMinutes();

  let totalEntryMinutes = entryHours * 60 + entryMinutes;
  let totalCurrentMinutes = currentHours * 60 + currentMinutes;

  let totalDuration = Math.floor(totalCurrentMinutes - totalEntryMinutes);

  let totalHours = Math.round(totalDuration / 60);
  let totalMinutes = totalDuration % 60;

  let totalCharge = 0;

  if (totalHours <= 2) {
    totalCharge = 10;
  } else if (totalMinutes > 0) {
    totalCharge = totalHours * 10;
  } else {
    totalCharge = (totalHours - 1) * 10;
  }

  return (
    <Container
      sx={{
        width: 300,
        height: 300,
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "column",
        border: "1px solid black",
        borderRadius: "5px",
        marginTop: "50px",
      }}
    >
      <Box sx={{ marginTop: "20px" }}>
        <Typography
          variant="body1"
          fontWeight="700"
          data-testid="deregister-car-registration"
        >
          Car Number : {exitSlotDetails?.carNumber}
        </Typography>
        <Typography variant="body1" fontWeight="700">
          Entry Time: {entryHours} : {entryMinutes}{" "}
          {entryHours > 12 ? "PM" : "AM"}
        </Typography>
        <Typography variant="body1" fontWeight="700">
          Exit Time: {currentHours}:{currentMinutes}{" "}
          {entryHours > 12 ? "PM" : "AM"}
        </Typography>
        <Typography
          variant="body1"
          fontWeight="700"
          data-testid="deregister-time-spent"
        >
          Total Duration: {totalDuration} minutes
        </Typography>
        <Typography
          variant="body1"
          fontWeight="700"
          data-testid="deregister-charge"
        >
          Amount To be Paid: {totalCharge}
        </Typography>
        <Button
          variant="contained"
          color="success"
          onClick={() => handlePayNow(exitSlotDetails?.carNumber)}
          data-testid="deregister-payment-button"
          sx={{ margin: "10px" }}
        >
          Pay Now
        </Button>
        <Button
          variant="contained"
          color="secondary"
          data-testid="deregister-back-button"
          onClick={() => navigate("/slots")}
        >
          Back
        </Button>
      </Box>
    </Container>
  );
};

export default ExitFromParking;
