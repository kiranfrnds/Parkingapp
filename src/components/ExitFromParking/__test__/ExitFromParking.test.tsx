/* eslint-disable jest/valid-expect-in-promise */
import { fireEvent, render, screen } from "@testing-library/react";
import { ParkingContext } from "../../../ParkingContext/ParkingContext";
import ExitFromParking from "../ExitFromParking";

const data = [
  {
    id: 1,
    carNumber: "KA-01-AA-1235",
    time: new Date(),
    available: false,
  },
];

const value = {
  parkingSlots: data,
  stateId: "1",
  setParkingSlots: jest.fn(),
  setStateId: jest.fn(),
};

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

beforeEach(() => {
  global.fetch = jest.fn(() => {
    return Promise.resolve({
      json: () =>
        Promise.resolve({
          status: 200,
          data: [
            {
              id: 1,
              carNumber: "KA-01-AA-1235",
              time: new Date(),
              available: false,
            },
          ],
        }),
    });
  }) as jest.Mock;
});

describe("ExitFromParking", () => {
  beforeAll(() => {
    jest.spyOn(window, "alert").mockImplementationOnce(() => {});
    jest.spyOn(console, "error").mockImplementationOnce(() => {});
  });

  it("snap shot test render", () => {
    const { container } = render(
      <ParkingContext.Provider value={value}>
        <ExitFromParking />
      </ParkingContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should make payment", () => {
    render(
      <ParkingContext.Provider value={value}>
        <ExitFromParking />
      </ParkingContext.Provider>
    );
    const button = screen.getByTestId("deregister-payment-button");
    fireEvent.click(button);
    expect(button).toBeDefined();
  });
  it("should goto back", () => {
    render(
      <ParkingContext.Provider value={value}>
        <ExitFromParking />
      </ParkingContext.Provider>
    );
    const button = screen.getByTestId("deregister-back-button");
    fireEvent.click(button);
    expect(button).toBeDefined();
  });
});
//////////////////////////////

// describe("ExitFromParking", () => {
//   it("should render", () => {
//     const { asFragment } = render(<ExitFromParking />);
//     expect(asFragment()).toMatchSnapshot();
//   });
//   it("should render total duration", () => {
//     render(<ExitFromParking />);
//     const totalDuration = screen.getByTestId("deregister-time-spent");
//     expect(totalDuration).toBeInTheDocument();
//   });
//   it("should render total charge", () => {
//     render(<ExitFromParking />);
//     const totalDuration = screen.getByTestId("deregister-charge");
//     expect(totalDuration).toBeInTheDocument();
//   });
//   it("should render payment button", () => {
//     render(<ExitFromParking />);
//     const button = screen.getByTestId("deregister-payment-button");
//     expect(button).toBeInTheDocument();
//   });
//   it("should render back button", () => {
//     render(<ExitFromParking />);
//     const button = screen.getByTestId("deregister-back-button");
//     expect(button).toBeInTheDocument();
//   });
//   it("should make post request on click of pay now button", () => {
//     render(<ExitFromParking />);
//     const button = screen.getByTestId("deregister-payment-button");
//     fireEvent.click(button);
//     expect(button).toBeInTheDocument();
//   });
//   it("should make goto slots page click on back button", () => {
//     render(<ExitFromParking />);
//     const button = screen.getByTestId("deregister-back-button");
//     fireEvent.click(button);
//     expect(button).toGetAttribute("href", "/slots");
//   });
// });
