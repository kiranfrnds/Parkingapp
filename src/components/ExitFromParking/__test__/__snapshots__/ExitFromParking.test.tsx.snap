// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExitFromParking snap shot test render 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1pbuu5h-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1xaekgw"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        data-testid="deregister-car-registration"
      >
        Car Number : 
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
      >
        Entry Time: 
         : 
         
        AM
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
      >
        Exit Time: 
        NaN
        :
         
        AM
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        data-testid="deregister-time-spent"
      >
        Total Duration: 
        NaN
         minutes
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        data-testid="deregister-charge"
      >
        Amount To be Paid: 
        NaN
      </p>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-bhgjed-MuiButtonBase-root-MuiButton-root"
        data-testid="deregister-payment-button"
        tabindex="0"
        type="button"
      >
        Pay Now
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        data-testid="deregister-back-button"
        tabindex="0"
        type="button"
      >
        Back
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
